website/blog/2024-03-20-snapshot-testing.md

```diff
@@ -0,0 +1,173 @@
+
+<LiteYouTubeEmbed
+    id="wQRGpWX3fsk"
+    title="WebdriverIO Tutorials: Snapshot Testing"
+/>
+
+Let's dive into each of these powerful capabilities.
+
+## DOM or Object Snapshots
+
+For evaluating the state of the DOM, a large object or the content of a UI element we often tend to copy the value into our test and manually update it if we change the behavior of our application or component.
+
+With text-based snapshots, we can just have this handled by WebdriverIO. For example, let's say we want to verify the state of our React component in a browser, we can just do the following:
+
+```tsx title="/src/component.test.tsx"
+import { expect, $ } from '@wdio/globals'
+import { render } from '@testing-library/react'
+
+function App() {
+    const [theme, setTheme] = useState('light')
+
+    const toggleTheme = () => {
+        const nextTheme = theme === 'light' ? 'dark' : 'light'
+        setTheme(nextTheme)
+    }
+
+    return <button onClick={toggleTheme}>
+        Current theme: {theme}
+    </button>
+}
+
+describe('React Component Testing', () => {
+    it('supports snapshot tests', async () => {
+        const { container } = render(<App />)
+        await expect(container).toMatchSnapshot()
+        await $('button').click()
+        await expect(container).toMatchSnapshot()
+    })
+})
+```
+
+WebdriverIO will automatically grab the DOM structure of the component and store a snapshot file in `/src/__snapshots__` directory next to your test.
+
+component.test.tsx.snap` with the following content:
+
+```
+// Snapshot v1
+
+exports[`React Component Testing > supports snapshot tests 1`] = `"<div><button>Current theme: light</button></div>"`;
+
+exports[`React Component Testing > supports snapshot tests 2`] = `"<div><button>Current theme: dark</button></div>"`;
+```
+
+If you prefer to keep the snapshots as part of your tests you can use `toMatchInlineSnapshot` instead:
+
+```tsx
+await expect(container).toMatchInlineSnapshot()
+```
+
+After running the test for the first time, WebdriverIO will make a change to the test and fill in the snapshot inline:
+
+```tsx
+await expect(container).toMatchInlineSnapshot(`"<div><button>Current theme: light</button></div>"`)
+```
+
+Now, if you make changes to your component that will impact all snapshots you can update them all in a single run by calling:
+
+```sh
+npx wdio run wdio.conf.ts --updateSnapshots
+# or
+npx wdio run wdio.conf.ts -s
+```
+
+This makes maintaining your tests so much easier. The same works for all other types of objects, e.g. CSS Properties, or the text content of an element. They all can be converted into a snapshot to simplify the assertion and keep your tests lean. This can also speed up your tests by merging many single assertions into one:
+
+```ts
+const elem = $('#alertBar')
+await expect(elem).toHaveAttribute('data-alert')
+await expect(elem).toHaveCSSProperty('background-color', 'green')
```

Comment:
This is not covered by the snapshot, we don't get a complete DOM snapshot with computed styles